pkg/grpc/server_test.go

```diff
@@ -91,6 +96,201 @@ func (a *APIServerSuite) Test_CustomAPI() {
+}
+
+func (s *pingServiceTestImpl) RegisterServiceHandler(ctx context.Context, mux *runtime.ServeMux, conn *grpc.ClientConn) error {
+	return v1.RegisterPingServiceHandler(ctx, mux, conn)
+}
+
+func (s *pingServiceTestImpl) AuthFuncOverride(ctx context.Context, fullMethodName string) (context.Context, error) {
+	return ctx, allow.Anonymous().Authorized(ctx, fullMethodName)
+}
+
+func (s *pingServiceTestImpl) Ping(context.Context, *v1.Empty) (*v1.PongMessage, error) {
+	s.RequestCount++
+
+	result := &v1.PongMessage{
+		Status: "test",
+	}
+
+	return result, nil
+}
+
+func (a *APIServerSuite) Test_Server_RateLimit_GRPC_Integration() {
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+
+	a.Run("no limiter", func() {
+		cfg := defaultConf()
+
+		api := NewAPI(cfg)
+		pingService := &pingServiceTestImpl{}
+		api.Register(pingService)
+		a.Assert().NoError(api.Start().Wait())
+		defer func() { api.Stop() }()
+
+		numOfRequests := 50
+		for i := 0; i < numOfRequests; i++ {
+			resp, err := http.Get("https://localhost:8080/v1/ping")
+			a.Require().NoError(err)
+			a.Require().Equal(http.StatusOK, resp.StatusCode)
+		}
+		a.Assert().Equal(numOfRequests, pingService.RequestCount)
+	})
+
+	a.Run("default unlimited", func() {
+		cfg := defaultConf()
+		cfg.RateLimiter = ratelimit.NewRateLimiter(0)
+
+		api := NewAPI(cfg)
+		pingService := &pingServiceTestImpl{}
+		api.Register(pingService)
+		a.Assert().NoError(api.Start().Wait())
+		defer func() { api.Stop() }()
+
+		numOfRequests := 50
+		for i := 0; i < numOfRequests; i++ {
+			resp, err := http.Get("https://localhost:8080/v1/ping")
+			a.Require().NoError(err)
+			a.Require().Equal(http.StatusOK, resp.StatusCode)
+		}
+		a.Assert().Equal(numOfRequests, pingService.RequestCount)
+	})
+
+	a.Run("hit rate limit", func() {
+		cfg := defaultConf()
+		cfg.RateLimiter = ratelimit.NewRateLimiter(3)
+
+		api := NewAPI(cfg)
+		pingService := &pingServiceTestImpl{}
+		api.Register(pingService)
+		a.Assert().NoError(api.Start().Wait())
+		defer func() { api.Stop() }()
+
+		hitLimit := false
+		requestCount := 0
+		for requestCount < 50 {
+			requestCount++
+			resp, err := http.Get("https://localhost:8080/v1/ping")
+			a.Require().NoError(err)
+
+			if requestCount <= 3 {
+				a.Require().Equal(http.StatusOK, resp.StatusCode)
+				continue
+			}
+
+			if resp.StatusCode != 200 {
+				a.Require().Equal(http.StatusTooManyRequests, resp.StatusCode)
+				hitLimit = true
+
+				// Request was rejected because of rate limit.
+				requestCount--
+				break
+			}
+		}
+		a.Assert().True(hitLimit)
+		a.Assert().Equal(requestCount, pingService.RequestCount)
+
+		// Wait for rate limit to refill.
+		time.Sleep(2 * time.Second)
+
+		requestCount++
+		resp, err := http.Get("https://localhost:8080/v1/ping")
+		a.Assert().NoError(err)
+		a.Assert().Equal(http.StatusOK, resp.StatusCode)
+
+		a.Assert().Equal(requestCount, pingService.RequestCount)
+	})
+}
+
```

Comment:
How about adding one more test that mix grpc and http interceptors as they share rate limiters